### DISCUSSION TITLE
v1.1.0 strange UNIQUE Index issue

### DISCUSSION THREAD

**Discussion by tartufella** [duckdb/duckdb#13900] [General]
Hello there,  I have a weird situation I'm not sure I understand.

The attached database has a table called Sales, it was generated through a .NET client using v1.1.0;

In the .NET code the following line is executed:
```sql
CREATE UNIQUE INDEX sales_idx ON Sales (CalendarID,ItemID,CountryID,ChannelDemandID,PriceType);
```

If I open up the creates database in the CLI/DBeaver and run the following I will see duplicates :
```sql
SELECT * FROM Sales WHERE CalendarID = 202526 AND ItemID = 550;
```

However if I also run : 
```sql
INSERT INTO Sales (CalendarID,itemID,CountryID,ChannelDemandID,PriceType) VALUES (202526,550,1,9,'PROMO');
```

This will fail as expected.  I'm not sure I understand how I can have both duplicates in the table at the same time that I cannot add duplicates.  I have also raised this same issue in the .NET repo.

If I revert the .NET version back to v1.0.2 then the UNIQUE INDEX works as expected but this still raises the issue of how a valid duckdb file can be generated that shows this issue.

Many thanks

//Marshall
[9cfb53b4-faec-4640-a783-d4c9a27ddf94.zip](https://github.com/user-attachments/files/16979048/9cfb53b4-faec-4640-a783-d4c9a27ddf94.zip)

**Comment by szarnyasg** [comment]
Hi @tartufella thanks for opening this ticket.

> In the .NET code the following line is executed:

Do you mean that it executes in .NET without returning an error?

In the CLI client, it fails as expected:

```console
D CREATE UNIQUE INDEX sales_idx ON Sales (CalendarID,ItemID,CountryID,ChannelDemandID,PriceType);
Constraint Error: Data contains duplicates on indexed column(s)
```

**Reply by tartufella** [reply]
Hi @szarnyasg yes, that's correct, it executes in .NET client without erroring and saves the file as provided.  The strange thing is that the database file is then in an inconsistent state where the index exists with duplicates in the table but you can not add new duplicates.

 SELECT * FROM duckdb_indexes shows that the index does exists as it was created by the .NET code but somehow the .NET code allowed duplicates into the table.  If I just revert the .NET client to v1.0.2 then it all works as expected.  I'm confused by this as I thought the .NET code simply proxies the queries into the duckdb library so it should show the same behaviour as the cli?

**Reply by Giorgi** [reply]
@tartufella Does the attached file already contains the index you added from .Net?

**Reply by tartufella** [reply]
Hi there, yes it does the attached file was created in .NET v1.1.0.  If I revert to v1.0.2 and run the same code there will be no duplicates in the table.

**Reply by Giorgi** [reply]
I responded to you in the .Net repo but do you insert data first or do you create an index first?

**Comment by Giorgi** [comment]
@tartufella @szarnyasg I took a look at the [repro](https://github.com/tartufella/DuckDBIssue) and this is what I found:

 There is no issue with index created on duplicated data. Index creation will fail if there is duplicate data in the table. But the issue is with the following:
 
 If you run statements from lines 11-15 you will end up with a `Sales` table and a unique index on the table. If you run the following query at that moment, you will see that there is no duplicated data in the `Sales` table:
```sql
SELECT * FROM Sales WHERE CalendarID = 202526 AND ItemID = 550 order by ChannelDemandId, PriceType;
```

Next, run the following statement:

```sql
INSERT OR IGNORE INTO Sales (CalendarID,ItemID,CountryID,ChannelDemandID,PriceType)
                SELECT *
                FROM
	                range(202505::INTEGER,202625::INTEGER) Calendar(CalendarID) 
                    CROSS JOIN (VALUES (103),(550),(9693)) Items(ItemID)
                    CROSS JOIN (VALUES (1)) Countries(CountryID)
                    CROSS JOIN (VALUES (3),(5),(7),(9)) Chanels(ChannelDemandID)
                    CROSS JOIN (SELECT DISTINCT PriceType FROM Sales) AS PriceTypes;
```

The insert statement will now insert duplicated rows in the table even though there is a unique index on the the table. You can verify it by running 

```sql
SELECT CalendarID,ItemID,CountryID,ChannelDemandID,PriceType 
            FROM Sales 
            WHERE 
                CalendarID = 202526 AND ItemID = 550 
            GROUP BY CalendarID,ItemID,CountryID,ChannelDemandID,PriceType
            HAVING COUNT(*) >1 ORDER BY ChannelDemandID, PriceType;
```

**Reply by Giorgi** [reply]
![image](https://github.com/user-attachments/assets/5793f7e3-11de-429d-a08d-cbf1d05a63f5)

As you can see after creating a unique index I can successfully insert duplicate rows.

@tartufella As this happens in the CLI, it's not related to .NET.

**Reply by szarnyasg** [reply]
Hi @Giorgi thanks for taking a look. I tried to reproduce this in the CLI on macOS with the script below but it does not reproduce for me. Does it occur for you in CLI or only in .NET?

```sql
CREATE TABLE SalesActuals AS SELECT * FROM read_parquet('sales_actuals.parquet');
CREATE TABLE SalesUsers AS SELECT * FROM read_parquet('sales_users.parquet');
CREATE TABLE Sales AS SELECT * FROM SalesActuals sa FULL JOIN SalesUsers users USING(CalendarID,ItemID,CountryID,ChannelDemandID,PriceType);
CREATE UNIQUE INDEX sales_idx ON Sales (CalendarID,ItemID,CountryID,ChannelDemandID,PriceType);
SELECT * FROM Sales WHERE CalendarID = 202526 AND ItemID = 550 order by ChannelDemandId, PriceType;
INSERT OR IGNORE INTO Sales (CalendarID,ItemID,CountryID,ChannelDemandID,PriceType)
                SELECT *
                FROM
	                range(202505::INTEGER,202625::INTEGER) Calendar(CalendarID) 
                    CROSS JOIN (VALUES (103),(550),(9693)) Items(ItemID)
                    CROSS JOIN (VALUES (1)) Countries(CountryID)
                    CROSS JOIN (VALUES (3),(5),(7),(9)) Chanels(ChannelDemandID)
                    CROSS JOIN (SELECT DISTINCT PriceType FROM Sales) AS PriceTypes;
SELECT CalendarID,ItemID,CountryID,ChannelDemandID,PriceType 
            FROM Sales 
            WHERE 
                CalendarID = 202526 AND ItemID = 550 
            GROUP BY CalendarID,ItemID,CountryID,ChannelDemandID,PriceType
            HAVING COUNT(*) >1 ORDER BY ChannelDemandID, PriceType;
```

**Reply by Giorgi** [reply]
See the screenshot above. It's from CLI on Windows that I installed earlier today with winget.

**Reply by szarnyasg** [reply]
Thanks! This definitely warrants investigation.

**Comment by Giorgi** [comment]
@tartufella Here is the issue: https://github.com/duckdb/duckdb/issues/13930

**Comment by taniabogatsch** [comment]
Hi @tartufella, we recently pushed a few fixes for our indexes here: https://github.com/duckdb/duckdb/pull/13858

After pushing these fixes, we cannot reproduce this issue in our latest `main` branch. Maybe you can double-check with the latest nightly. I'll close this discussion as solved for now, but should you run into more issues, please go ahead and open a bug report so we can look into them.

**Reply by tartufella** [reply]
Thanks very much @taniabogatsch the issue seems to have resolved itself certainly with v1.01.4917.

thanks also @Giorgi, will wait to upgrade until a new release is made with this fixed rolled in.

**Comment by Giorgi** [comment]
@tartufella Just pushed v1.1.1 to NuGet that includes a fix for this issue.
